### PR TITLE
cli: new command `auth-session {login,logout,list}`

### LINKED ISSUE #43870: security: need a license for HTTP access  #42567
**tldr: the security fix for  #42567 has been backported to 2.1, 19.1 and 19.2, however it also breaks legitimate use cases. We need to find a workaround.**

(CockroachCloud users are not affected.)

### Background

Issue  #42567 describes how previous versions of CockroachDB did not authorize accesses to the status endpoints including e.g. `/admin/_v1/logs` so any non-root user could access it.

This security vulnerability has been patched by requiring all such access to be authenticated as an admin user.

### Description of the problem

tldr: without an Enterprise license, the only admin user is `root` and `root` is not authorized to use the HTTP endpoints. **Therefore, core users without a valid license cannot use advanced HTTP monitoring any more.** This is a regression.

Details:

- Only users with a valid password are allowed to log in to the privileged HTTP endpoints, using their password to create a valid HTTP auth cookie via `/login` (This is a product feature.) CockroachDB does not yet support client authentication using client TLS certificates.

- Meanwhile, `root` is prevented from having a password (This is a current product feature, discussed in #43847)

- Without an enterprise license, only `root` is an admin user: An enterprise license is required to make other users members of the `admin` role.

- This does not affect CockroachCloud customers because all SQL users created with the CC console are admins.

### Directions for workaround and solutions

Any of the following product directions would un-block HTTP monitoring on clusters without a license:

1. create a `cockroach` CLI command that creates/obtains a web cookie for a given user. The CLI command would use a client cert to obtain this cookie. Prototype here:  #43872

2. document how to manually create an auth cookie for the `root` user in `system.web_sessions` using a custom client program using SQL (possibly providing a Python script)

3. pre-define another user than `root` that is also an admin in every cluster (e.g. create an `operator` user using a cluster migration), without requiring a license. This was proposed by @israellot in https://github.com/cockroachdb/cockroach/issues/43847#issuecomment-572806638

4. enabling `root` to have a password #43847; this in turn would enable a root client to obtain an auth cookie via the `/login` endpoint then use that to perform further monitoring.

5. enabling authentication using TLS certs in addition to web cookies, so that a client can perform HTTP monitoring using the regular root client cert.

6. open a separate HTTP port for insecure (non-authenticated) status/debug endpoints, and let the DBA/sysadm manage security at the network level (i.e. with firewalling)

| Solution | Complexity | Suitable for backport in 19.2/19.1/2.1 |
|----------|------------|----------------------------------------|
| 1 - new CLI command  #43872 | small/medium | yes |
| 2 - document manual cookie creation | medium | yes
| 3 - predefined operator user | small | unsure/yes |
| 4 - root password #43847 #43893 | small | unsure/no |
| 5 - http auth via client TLS certs | medium | no |
| 6 - separate insecure http port | small | unsure/no |

(we can add more solution design points to this list)

### PR DESCRIPTION
**Pull request by knz**
Fixes #43870.

tldr: this adds new CLI commands to log users in and out of the
HTTP interface and produce a HTTP cookie for use in monitoring
scripts. This is suitable for use by the `root` user without an
Enterprise license.

Also the new feature is client-side only, so the client binary with
this feature can be used with a CockroachDB server/cluster running at
an older version.

**Motivation:** users who wish to use certain HTTP monitoring tools,
in particular those that retrieve privileged information like logs,
need a valid HTTP authentication token for an admin user (#42567). This token
can be constructed by accessing the HTTP endpoint `/login`, however:

- manually crafting the token using `/login` is cumbersome;
- it's not possible to use `/login` for the `root` user (#43847);
- it's not possible to create another admin user than `root` without
  a valid Enterprise license (because that requires role management).

**Solution:**

```
cockroach auth-session login <username> [--expire-after=...] [--only-cookie]
cockroach auth-session logout <username>
cockroach auth-session list
```

- all three commands also support the standard SQL command-line
  arguments, e.g. `--url`, `--certs-dir`, `--echo-sql` and
  `--format`.
- the `--expire-after` argument customizes the expiry period. The
  default is one hour.
- the `--only-cookie` arguments limits the output of the command
  to just the HTTP cookie. By default, the session ID and
  the authentication cookie are printed using regular table formatting.

Also see the two release notes below.

Release note (cli change): Three new CLI commands `cockroach
auth-session login`, `cockroach auth-session list` and `cockroach
auth-session logout` are now provided to facilitate the management of
web sessions. The command `auth-session login` also produces a HTTP
cookie which can be used by non-interactive HTTP-based database
management tools. It also can generate such a cookie for the `root`
user, who would not otherwise be able to do so using a web browser.

Release note (security update): The new command `cockroach
auth-session login` (reserved to administrators) is able to create
authentication tokens with an arbitrary expiration date. Operators
should be careful to monitor `system.web_sessions` and enforce
policy-mandated expirations either using SQL queries or the new
command `cockroach auth-session logout`.

### CODE REVIEW

**Comment by knz** [comment]
This change has zero impact on the stability of the server side and is suitable for backporting to 19.2, 19.1 and 2.1.

**Comment by aaron-crl** [comment]
I don't have depth of knowledge in the codebase to review for implementation correctness yet so we should probably loop in another engineer for that.

(1) What functionality is available given an auth token (does this expose more than metrics/logging)?
(2) How hard would it be to delegate access to metrics and logging to another role/user in non-enterprise clusters?

**Comment by knz** [comment]
> I don't have depth of knowledge in the codebase to review for implementation correctness yet so we should probably loop in another engineer for that.

:+1: I trust @ajwerner can help

> (1) What functionality is available given an auth token (does this expose more than metrics/logging)?

It gives access to the web UI and all that you can do via the HTTP endpoint using that user's credentials (mostly defined via its SQL privileges and admin role membership).

(I think that inventorizing what is (or is not) available for UI users is out of scope of this PR? WDYT? Unless you were asking out of general curiosity)

> (2) How hard would it be to delegate access to metrics and logging to another role/user in non-enterprise clusters?

- metrics: not hard, although I believe they are not restricted today? (I can check)
- logs: insanely hard -- 50-80% of our logging information can potentially leak sensitive data, including user credentials, and introducing infrastructure to enforce sensitive field redaction throughout our source code would be a huge task.

Also here again I would be sympathetic to an effort to pre-define an "operator" role/user as a feature, but maybe that would be a different/orthogonal issue?

**Comment by aaron-crl** [comment]
I think the ability for administrators to revoke sessions is a great feature add.

Aside from that, this doesn't seem like it functionally changes the security controls in place but it does make it more likely that a credential will be used it ways that might expose it to an attacker.

(3) Given that this token is analogous to a web user's valid credentials, what logging events or other mitigating controls (if any) would improve the ability to detect or stop token misuse?

(4) Re 2: Would it be possible to define a read-only log consumer role to prepare for a point in the future where sanitized or more granular logging is possible? If so, let's create another issue to track and prioritize that.

**Review by ajwerner — COMMENTED** [review]
Other than using the flag you added :lgtm:

> Reviewed 6 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43872#-:-LyZhPHO3le0LJMK6XG_:b2zqm5o)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aaron-crl, @ajwerner, and @knz)

___
*[pkg/cli/auth.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43872#-LyZh4xdAyDD_zeaIh0f:-LyZh4xdAyDD_zeaIh0g:b-ao4epe) ([raw file](https://github.com/cockroachdb/cockroach/blob/42228ddd9ada4cc8c14d29d84e15a9bca2e699cc/pkg/cli/auth.go#L57)):*
> ```Go
> 	}
> 
> 	expiration := timeutil.Now().Add(1 * time.Hour)
> ```

should this use `authCtx.validityPeriod`?
___
*[pkg/cli/auth.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43872#-LyZgu1LAbE4fFD2oiX9:-LyZgfDp0jSXcOtxrYSE:b-1ktvd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/42228ddd9ada4cc8c14d29d84e15a9bca2e699cc/pkg/cli/auth.go#L59)):*
> ```Go
> 	expiration := timeutil.Now().Add(1 * time.Hour)
> 
> 	insertSessionStmt := `
> ```

Would it make sense to verify that a user with the name `username` exists?

**Review by knz — COMMENTED** [review]
> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43872)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aaron-crl and @ajwerner)

___
*[pkg/cli/auth.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43872#-LyZh4xdAyDD_zeaIh0f:-Lya3Z-FBkrxXQq14JvI:b-vqls7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/42228ddd9ada4cc8c14d29d84e15a9bca2e699cc/pkg/cli/auth.go#L57)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should this use `authCtx.validityPeriod`?
</blockquote></details>

:man_facepalming: yes. Fixed.
___
*[pkg/cli/auth.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43872#-LyZgu1LAbE4fFD2oiX9:-Lya3bvEAwxEHbY6Ukmo:b26gnw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/42228ddd9ada4cc8c14d29d84e15a9bca2e699cc/pkg/cli/auth.go#L59)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Would it make sense to verify that a user with the name `username` exists?
</blockquote></details>

I guess it would make sense. Added.

**Comment by knz** [comment]
bors r=ajwerner,aaron-crl

**Comment by knz** [comment]
oops typo

bors r-

**Comment by knz** [comment]
bors r=ajwerner,aaron-crl
